lib/Parse/Parser.cpp

```diff
@@ -709,6 +709,23 @@ void Parser::skipUntilConditionalBlockClose() {
   }
 }
 
+bool Parser::loadCurrentSyntaxNodeFromCache() {
+  // Don't do a cache lookup when not building a syntax tree since otherwise
+  // the corresponding AST nodes do not get created
+  if (!SF.shouldBuildSyntaxTree()) {
+    return false;
+  }
+  unsigned LexerOffset =
+      SourceMgr.getLocOffsetInBuffer(Tok.getLoc(), L->getBufferID());
+  if (auto TextLength = SyntaxContext->loadFromCache(
+          LexerOffset - LeadingTrivia.getTextLength())) {
+    L->advanceTo(Tok, LeadingTrivia, TextLength);
+    L->lex(Tok, LeadingTrivia, TrailingTrivia);
+    return true;
+  }
```

Comment:
I feel this needlessly complicated.
How about:
```cpp
  unsigned Offset = SourceMgr.getLocOffsetInBuffer(Tok.getLoc(), L->getBufferID());
  Offset -= LeadingTrivia.getTextLength();
  if (auto TextLength = SyntaxContext->loadFromCache(Offset)) {
    L->resetToOffset(Offset + TextLength);
    L->lex(Tok, LeadingTrivia, TrailingTrivia);
    return true;
  }
```
where `Lexer:: resetToOffset` is just:
```swift
Lexer::resetToOffset(size_t Offset) {
  assert(BufferStart + Offset <= BufferEnd);
  CurPtr = BufferStart + Offset;

  lexImpl();
}
```